### PR TITLE
THE 2023 release

### PR DESCRIPTION
**Pull request by RedLeader962**
For approval by @pomerlef 

Preparation for release version with commits spanning over close to 5 years.

Note: the code diff between the `release` branch and the `master` branch wont reflect the breath of commit added to the proposed release version both in number and type. 

~~Even though there's no API breaking change, since there was a lot of major change and improvement, I propose that this release bump the repository version from to `1.3.1` →  `2.0.0`.~~
After discussion, will keep major no to 1 for the release

In this PR:
- Doc update
- changelog update
- fixed codebase install instabilities issues [[#538](https://github.com/norlab-ulaval/libpointmatcher/pull/538) and [#539](https://github.com/norlab-ulaval/libpointmatcher/pull/539)]
- added submodule [norlab-shell-script-tools](https://github.com/norlab-ulaval/norlab-shell-script-tools)
- added submodule [norlab-build-system](https://github.com/norlab-ulaval/norlab-build-system)
- refactor out part of the *libpointmatcher-build-system* logic to [norlab-build-system](https://github.com/norlab-ulaval/norlab-build-system) library [[#545](https://github.com/norlab-ulaval/libpointmatcher/pull/545)]
- **a LOT of other codebase fixes and improvements**

### CODE REVIEW

**Comment by RedLeader962** [comment]
@pomerlef After discussion with @boxanm, we realized we need to make some changes. Just switched the PR to draft.

**Comment by RedLeader962** [comment]
@pomerlef Also, @boxanm make a good argument about the version bump

> I don't think the release tag should be 2.0, since we still keep backwards compatibility with 1.* versions. I'd go with 1.4 instead and use 2.0 next time, after we migrate to C++17/20 and new release of libnabo .

I dont mind the version number, however we are dropping support for two EOL ubuntu version in this release (Ubuntu 12.04 and 14.04). So we need to discuss this

**Comment by pomerlef** [comment]
I'm more aligned with keeping 1.* until we break the API

**Comment by RedLeader962** [comment]
I've merged some fix and improvement to  NBS as well as its integration to LPM

**Comment by RedLeader962** [comment]
I've merged PR #546 in the release. Those are mainly fix and stabilisation improvement related to the build-system and additions to the repository management:

#### Build-system related
- I've added tree scripts to easily launch the build system. They are designed to be used locally as part of a development workflow on a workstation and as part of a CI infrastructure such as TeamCity or GitHub action. See `lpm_crawl_[dependencies|libpointmatcher]_build_matrix[.bleeding].bash` in [`build_system/`](https://github.com/norlab-ulaval/libpointmatcher/blob/6ae246eb488420dfd02687023fce9b731622a0fc/build_system) directory. 
- I've added improvements and fixes to the main `README.md` header and the _Contributing_ instructions on how to use the _libpointmatcher-build-system_ localy   
- Most `libpointmatcher-build-system` related script are now properly tested using [norlab-shell-script-tools (N2ST)](https://github.com/norlab-ulaval/norlab-shell-script-tools) testing tools
 
#### Repository management related 
- I've added a pull request template with full instructions on how to proceed for submitting a PR, 
- improved the `CODEOWNERS `file so that key person get a review request notification when certain area of the codebase are modified by a pull request, eg: if someone push a PR that mess with the `build_system/` directory I'll get notified

**Comment by boxanm** [comment]
Before release, I'd like to merge the PR that's supposed to turn up from https://github.com/norlab-ulaval/libpointmatcher/issues/454 . Since we dropped support for versions older than Ubuntu 18.04, we could also merge https://github.com/norlab-ulaval/libpointmatcher/pull/536 .

**Comment by RedLeader962** [comment]
... and I've just realized I accidentally merged PR #540 from `develop` in the `release` branch, let me know @boxanm I you  did not plan to make it part of the this release, I'll find a way to rewrite the commit history

**Comment by RedLeader962** [comment]
No stress @boxanm, I think you can take the time that you need. People have been waiting for a new release for a few years, they can wait for a few more days. I've notified those that were experiencing install/compiling issue to use the `develop` branch in the meantime.
Note: I'm switching to integrating `NBS` to _libnabo_ and _norlab-icp-mapper_ codebase

**Comment by RedLeader962** [comment]
@pomerlef I've also added instruction and references to gently guide contributor toward adopting [Conventional Commit](https://www.conventionalcommits.org) so that we can later use semantic-versioning to automate release process. It's too soon to enforce it's uses since libpointmatcher contributors already have an establish workflow and need some kinds of heads up. See [`pull_request_template.md`](https://github.com/norlab-ulaval/libpointmatcher/blob/6ae246eb488420dfd02687023fce9b731622a0fc/.github/pull_request_template.md) and [`commit_msg_reference.md`](https://github.com/norlab-ulaval/libpointmatcher/blob/6ae246eb488420dfd02687023fce9b731622a0fc/commit_msg_reference.md)

File: .github/CODEOWNERS
Comment (pomerlef) [inline .github/CODEOWNERS:21]: Not sure what is the purpose of this file, but this list doesn't list ok complet

**Review by pomerlef — CHANGES_REQUESTED** [review]

File: .github/CODEOWNERS
Comment (RedLeader962) [reply .github/CODEOWNERS:21]: @pomerlef The `CODEOWNERS` file designate who can approve a pull request on a protected branch for a given file/directories. They will also be notified when change are performed on those file/directories eg any change in the `pointmatcher` directory and child will trigger a code review request to the designated code owner.
I've use the name of those that had performed code review recently as a starting point.

**Comment by RedLeader962** [comment]
@pomerlef I was wondering if we should put the new `Contributing` section currently in the main readme to a dedicated file eg `docs/contributing.md`.
Also I forgot to mention the new installer script in the readme. I'll add it after the holiday

**Review by pomerlef — APPROVED** [review]

**Pull request merged**
